vault/resource_database_secrets_mount.go

```diff
@@ -0,0 +1,334 @@
+				"Must be prefixed with the name of one of the supported database engine types.",
+			ValidateFunc: func(i interface{}, s string) ([]string, []error) {
+				var errs []error
+				v, ok := i.(string)
+				if !ok {
+					errs = append(errs, fmt.Errorf("expected type of %q to be string", s))
+				} else if err := validateDBPluginName(v); err != nil {
+					errs = append(errs, err)
+				}
+				return nil, errs
+			},
+		},
+		"verify_connection": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: "Specifies if the connection is verified during initial configuration.",
+			Default:     true,
+		},
+		"allowed_roles": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "A list of roles that are allowed to use this connection.",
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"root_rotation_statements": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "A list of database statements to be executed to rotate the root user's credentials.",
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
+		"data": {
+			Type:        schema.TypeMap,
+			Optional:    true,
+			Description: "A map of sensitive data to pass to the endpoint. Useful for templated connection strings.",
+			// TODO: revert to true
+			Sensitive: false,
+		},
+	}
+}
+
+func setCommonDatabaseSchema(s schemaMap) schemaMap {
+	for k, v := range getCommonDatabaseSchema() {
+		s[k] = v
+	}
+	return s
+}
+
+func databaseSecretsMountCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	var root string
+	if d.IsNewResource() {
+		root = d.Get("path").(string)
+		if err := createMount(d, client, root, "database"); err != nil {
+			return err
+		}
+	} else {
+		if err := mountUpdate(d, meta); err != nil {
+			return err
+		}
+		root = d.Id()
+	}
+	d.SetId(root)
+
+	var count int
+	seen := make(map[string]bool)
+	for _, engine := range dbEngines {
+		if v, ok := d.GetOk(engine.Name()); ok {
+			for i := range v.([]interface{}) {
+				prefix := engine.ResourcePrefix(i)
+				name := d.Get(prefix + "name").(string)
+				path := databaseSecretBackendConnectionPath(root, name)
+				if _, ok := seen[name]; ok {
+					return fmt.Errorf("duplicate name %q for engine %#v", name, engine)
+				}
+				seen[name] = true
+				if err := writeDatabaseSecretConfig(d, client, engine, i, true, path); err != nil {
+					return err
+				}
+				count++
+			}
+		}
+	}
+
+	if err := databaseSecretsMountRead(d, meta); err != nil {
+		return err
+	}
+
+	action := "Created"
+	if !d.IsNewResource() {
+		action = "Updated"
+	}
+
+	log.Printf("[DEBUG] %s %d database connections under %q", action, count, root)
+	return nil
+}
+
+func databaseSecretsMountRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	if err := readMount(d, meta, true); err != nil {
+		return err
+	}
+
+	root := d.Id()
+	if root == "" {
```

Comment:
Wondering why this is not an error. If we're in the Read func, should there be a case where the ID on the resource isn't set?

Reply:
Yeah, sort of misleading. Essentially `readMount()` takes care of setting to `""` if the mount was not found.

Reply:
Added a comment.

Reply:
Thanks!